helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/LeastPartitionCountConstraint.java

```diff
@@ -0,0 +1,55 @@
+package org.apache.helix.controller.rebalancer.waged.constraints;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import static java.lang.Math.max;
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+
+/**
+ * Evaluate the proposed assignment according to the instance's partition count.
+ */
+class LeastPartitionCountConstraint extends SoftConstraint {
+  static LeastPartitionCountConstraint INSTANCE  = new LeastPartitionCountConstraint();
```

Comment:
Singleton instance has a better performance of GC. The downside is we lost the ability to customize the constraint of different score ranges, but I don't think we would need the requirement. I don't think there's any concurrent issue related to the class if it's singleton, can you specify?

Reply:
By concurrent issue, I mean if you need to keep some internal variable as a field in the constraint object, then we will need to do concurrent control. Because we might have multiple callers. I'm not saying it would be the case, just a thought.

Actually, the score range is a good point. We should have a consistent range. Given that saying, singleton could help.